### PR TITLE
samples: sensor: Add generic accelerometer polling sample

### PR DESCRIPTION
**Pull request by soburi**
Add a generic polling sample for 3-Axis accelerometers.
This sample demonstrates how to data fetch and print to the console.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

-------------------------------------------

Re-create PR. Previously created one is already close.
https://github.com/zephyrproject-rtos/zephyr/pull/45034

### CODE REVIEW

**Comment by ycsin** [comment]
I think you need to specify a board and an overlay for it?
https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/sensor/adxl362

File: samples/sensor/adxl345/Kconfig [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2022 TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>
+# SPDX-License-Identifier: Apache-2.0
+
+# Workaround for not being able to have commas in macro arguments
+DT_COMPAT_ADI_ADXL345 := adi,adxl345
```
Comment (galak) [inline samples/sensor/adxl345/Kconfig:5]: This should be with the sensor driver and not here.

File: samples/sensor/adxl345/src/main.c [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+#include <device.h>
+#include <drivers/sensor.h>
+
+void main(void)
+{
+	struct sensor_value accel[3];
+
+	const struct device *dev = DEVICE_DT_GET_ANY(adi_adxl345);
+
+	if (dev == NULL) {
```
Comment (galak) [inline samples/sensor/adxl345/src/main.c:17]: dev will never be null,  replace this check with a `device_is_ready(dev)` check.  See other samples/sensors for example.

**Review by galak — COMMENTED** [review]
Looking at main.c, not sure if this adds much beyond what the sensor shell can do.

**Comment by soburi** [comment]
@MaureenHelm , @galak 
> This is very close to being a generic accelerometer sample. Can you look at doing something like #47005 to make this work for more than just one accelerometer driver?

> Looking at main.c, not sure if this adds much beyond what the sensor shell can do.


I got it. I'll rewrite this PR as generic  accelerometer sample

**Comment by soburi** [comment]
@galak @MaureenHelm 

Rewrite this PR as generic accelerometer sample like as #47005.

File: boards/arm/96b_wistrio/96b_wistrio.dts [hunk 3]
```diff
@@ -93,8 +94,8 @@
 	pinctrl-names = "default";
 	status = "okay";
 
-	lis3dh@32 {
-		compatible = "st,lis3dh";
+	lis3dh: lis3dh@32 {
+		compatible = "st,lis3dh", "st,lis2dh";
```
Comment (galak) [inline boards/arm/96b_wistrio/96b_wistrio.dts:98]: why adding `"st,lis2dh"` here?

File: drivers/sensor/adxl345/Kconfig [hunk 4]
```diff
@@ -3,10 +3,13 @@
...
 # SPDX-License-Identifier: Apache-2.0
 
+DT_COMPAT_ADXL345 := adi,adxl345
+
 config ADXL345
 	bool "ADXL345 Three Axis accelerometer"
 	default y
 	depends on DT_HAS_ADI_ADXL345_ENABLED
 	depends on I2C || SPI
+	default $(dt_compat_enabled,$(DT_COMPAT_ADXL345))
```
Comment (galak) [inline drivers/sensor/adxl345/Kconfig:13]: Not needed, this is happening with the `default y` and `depends on DT_HAS_ADI_ADXL345_ENABLED`.  Same comment for all the sensor Kconfig changes.

**Review by galak — CHANGES_REQUESTED** [review]
The driver/sensor/Kconfig changes are not needed and should be removed.

File: boards/arm/bbc_microbit/bbc_microbit.dts [hunk 5]
```diff
@@ -17,6 +17,7 @@
 		sw0 = &buttonA;
 		sw1 = &buttonB;
 		magn0 = &lsm303agr_magn;
+		accel0 = &mma8653fc;
 		watchdog0 = &wdt0;
```
Comment (galak) [inline boards/arm/bbc_microbit/bbc_microbit.dts:21]: accel1?

File: samples/sensor/accel_polling/sample.yaml [hunk 6]
```diff
@@ -0,0 +1,50 @@
+sample:
+  name: Accelerometer polling sample
+tests:
+  sample.sensor.accel_polling:
+    harness: console
+    tags: sensors
+    filter: dt_alias_exists("accel0")
+    platform_exclude: sensortile_box thingy53_nrf5340_cpuapp lpcxpresso55s69_cpu1
```
Comment (galak) [inline samples/sensor/accel_polling/sample.yaml:8]: why are these particular platforms excluded?

**Review by galak — COMMENTED** [review]
Will leave it to @MaureenHelm if we should remove the samples this replaces in this PR or as a follow up.

File: boards/arm/bbc_microbit/bbc_microbit.dts (hunk 5 above)
Comment (soburi) [reply boards/arm/bbc_microbit/bbc_microbit.dts:21]: There are some variations of the BBC microbit. Each has one accelerometer, but the type of sensor is different (LSM303AGR or MMA8653FC).

The dts enables MMA8653FC sensor, and other sensors are disabled.
So, I set mma8653fc to accel0 and no use lsm303agr.

(I think it is intend to enable/disable each sensors by overlay file.)

Comment (soburi) [reply boards/arm/bbc_microbit/bbc_microbit.dts:21]: There are some variations of the BBC microbit. Each has one accelerometer, but the type of sensor is different (LSM303AGR or MMA8653FC).

The dts enables MMA8653FC sensor, and other sensors are disabled.
So, I set mma8653fc to accel0 and no use lsm303agr.

(I think it is intend to enable/disable each sensors by overlay file.)

File: drivers/sensor/adxl345/Kconfig (hunk 4 above)
Comment (soburi) [reply drivers/sensor/adxl345/Kconfig:13]: Removed.

File: boards/arm/96b_wistrio/96b_wistrio.dts (hunk 3 above)
Comment (soburi) [reply boards/arm/96b_wistrio/96b_wistrio.dts:98]: LIS3DH uses LIS2DH driver.
Enabling LIS2DH driver need to set DT_HAS_ST_LIS2DH_ENABLED.
By adding "st,lis2dh" to the compatibile, this option turn on.

If only "st,lis3dh" exists in the compatible option, 
only enable DT_HAS_ST_LIS3DH_ENABLED, but DT_HAS_ST_LIS2DH_ENABLED is not enabled.

**Review by galak — DISMISSED** [review]

File: boards/arm/actinius_icarus/Kconfig.defconfig [hunk 7]
```diff
@@ -8,6 +8,13 @@ if BOARD_ACTINIUS_ICARUS || BOARD_ACTINIUS_ICARUS_NS
 config BOARD
 	default "actinius_icarus"
 
+if SENSOR
+
+config I2C
+	default y
+
+endif # SENSOR
```
Comment (gmarull) [inline boards/arm/actinius_icarus/Kconfig.defconfig:16]:
```
config I2C
	default y if SENSOR
```

File: boards/arm/actinius_icarus_bee/Kconfig.defconfig [hunk 8]
```diff
@@ -8,6 +8,13 @@ if BOARD_ACTINIUS_ICARUS_BEE || BOARD_ACTINIUS_ICARUS_BEE_NS
 config BOARD
 	default "actinius_icarus_bee"
 
+if SENSOR
+
+config I2C
+	default y
+
+endif # SENSOR
```
Comment (gmarull) [inline boards/arm/actinius_icarus_bee/Kconfig.defconfig:16]: ditto (same for all others)

File: samples/sensor/accel_polling/README.rst [hunk 9]
```diff
@@ -0,0 +1,38 @@
+.. _accel_polling:
+
+Generic 3-Axis accelerometer polling sample
```
Comment (gmarull) [inline samples/sensor/accel_polling/README.rst:3]: for bisectability: sample should be the last addition

File: samples/sensor/accel_polling/sample.yaml [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    tags: sensors
+    platform_allow: sensortile_box
+    harness_config:
+      type: multi_line
+      regex:
+        - "^\\s*lis2dw12@0 \\[m\/s\\^2\\]:    \\(\\s*-?[0-9\\.]*,\\s*-?[0-9\\.]*,\\s*-?[0-9\\.]*\\)$"
+        - "^\\s*lsm6dso@1 \\[m\/s\\^2\\]:    \\(\\s*-?[0-9\\.]*,\\s*-?[0-9\\.]*,\\s*-?[0-9\\.]*\\)$"
+        - "^\\s*iis3dhhc@2 \\[m\/s\\^2\\]:    \\(\\s*-?[0-9\\.]*,\\s*-?[0-9\\.]*,\\s*-?[0-9\\.]*\\)$"
+    integration_platforms:
+      - sensortile_box                  # lis2dw12, lsm6dso, iisdhhc
```
Comment (gmarull) [inline samples/sensor/accel_polling/sample.yaml:38]: IMO this overcomplicates the sample.yaml, the generic entry should be sufficient for all cases even if not perfect, in the end we're validating that something is printed...

File: samples/sensor/accel_polling/src/main.c [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+};
+
+void format_sensor_value(struct sensor_value *value, char *buf, size_t len)
+{
+	if (value->val1 == 0) {
+		snprintf(buf, len, "  %c0.%06u", (value->val2 < 0 ? '-' : ' '), abs(value->val2));
+		return;
+	}
+	snprintf(buf, len, "%4d.%06u", value->val1, abs(value->val2));
+}
```
Comment (gmarull) [inline samples/sensor/accel_polling/src/main.c:59]: why not use `sensor_value_to_double` and print the value?

File: samples/sensor/accel_polling/src/main.c [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+		}
+	}
+
+	while (1) {
+		for (uint32_t i = 0; i < ARRAY_SIZE(sensors); i++) {
+			ret = print_accels(sensors[i]);
+			if (ret < 0) {
+				return;
+			}
+			k_sleep(K_MSEC(1000/ARRAY_SIZE(sensors)));
```
Comment (gmarull) [inline samples/sensor/accel_polling/src/main.c:106]: why add sleep between different accelerometer reads and not a `k_sleep(K_MSEC(1000))` after all are processed?

File: samples/sensor/accel_polling/src/main.c [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+#if DT_NODE_EXISTS(DT_ALIAS(accel7))
+	DEVICE_DT_GET(DT_ALIAS(accel7)),
+#endif
+#if DT_NODE_EXISTS(DT_ALIAS(accel8))
+	DEVICE_DT_GET(DT_ALIAS(accel8)),
+#endif
+#if DT_NODE_EXISTS(DT_ALIAS(accel9))
+	DEVICE_DT_GET(DT_ALIAS(accel9)),
+#endif
+};
```
Comment (gmarull) [inline samples/sensor/accel_polling/src/main.c:44]: hardcoding 10 instances this way doesn't look great. First, do we have any board with 10 accelerometers? Second, why not use something like `LISTIFY` to make this more compact?

Maybe a Kconfig option could actually set the number of accelerometers to expect, to be adjusted in a board overlay if > 1.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: samples/sensor/accel_polling/src/main.c [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+ * Copyright (c) 2022 TOKITA Hiroshi <tokita.hiroshi@fujitsu.com
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <drivers/sensor.h>
+#include <zephyr.h>
+#include <stdlib.h>
+#include <stdio.h>
```
[the remark this answers is not in the document]
Comment (soburi) [reply samples/sensor/accel_polling/src/main.c]: Fixed it.

File: boards/arm/actinius_icarus/Kconfig.defconfig (hunk 7 above)
Comment (soburi) [reply boards/arm/actinius_icarus/Kconfig.defconfig]: Fixed it.

File: boards/arm/actinius_icarus_bee/Kconfig.defconfig (hunk 8 above)
Comment (soburi) [reply boards/arm/actinius_icarus_bee/Kconfig.defconfig]: Fixed it, and also fixes for all similars.

File: samples/sensor/accel_polling/README.rst (hunk 9 above)
Comment (soburi) [reply samples/sensor/accel_polling/README.rst:3]: I reordered commits.

File: samples/sensor/accel_polling/sample.yaml (hunk 10 above)
Comment (soburi) [reply samples/sensor/accel_polling/sample.yaml]: Removing board specific cases.

File: samples/sensor/accel_polling/src/main.c (hunk 11 above)
Comment (soburi) [reply samples/sensor/accel_polling/src/main.c]: No strong reason is there.
Fix to use sensor_value_to_double.

File: samples/sensor/accel_polling/src/main.c (hunk 12 above)
Comment (soburi) [reply samples/sensor/accel_polling/src/main.c]: No strong reason is there.
Fix to fetch data from all sensors wit 1s interval.

File: samples/sensor/accel_polling/src/main.c (hunk 13 above)
Comment (soburi) [reply samples/sensor/accel_polling/src/main.c]: Rewrite with LISTIFY macro.  (I didn't know the existence of this macro. It is very helpful!)
As a result of the change, defining an alias for accelX will automatically work with this sample.
I think it's easier than defining a maximum in Kconfig because all you have to do is add an alias in the overlay.

How about this change?

File: samples/sensor/accel_polling/src/main.c [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+	}
+
+	while (1) {
+		for (uint32_t i = 0; i < ARRAY_SIZE(sensors); i++) {
+			ret = print_accels(sensors[i]);
+			if (ret < 0) {
+				return;
+			}
+		}
+		k_sleep(K_MSEC(1000));
```
Comment (ycsin) [inline samples/sensor/accel_polling/src/main.c:67]: could have just used [`k_msleep`](https://docs.zephyrproject.org/2.7.0/reference/kernel/threads/index.html#c.k_msleep)

Comment (soburi) [reply samples/sensor/accel_polling/src/main.c]: Changed it to k_msleep.

File: samples/sensor/accel_polling/sample.yaml [hunk 16]
```diff
@@ -0,0 +1,27 @@
+sample:
+  name: Accelerometer polling sample
+tests:
+  sample.sensor.accel_polling:
+    harness: console
+    tags: sensors
+    filter: dt_alias_exists("accel0")
+    platform_exclude: sensortile_box thingy53_nrf5340_cpuapp lpcxpresso55s69_cpu1
```
[the remark this answers is not in the document]
Comment (soburi) [reply samples/sensor/accel_polling/sample.yaml]: I forgot delete it.
This line deleted now.

File: samples/sensor/accel_polling/src/main.c [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+	       sensor_value_to_double(&accel[2]));
+
+	return 0;
+}
+
+void main(void)
+{
+	int ret;
+
+	for (uint32_t i = 0; i < ARRAY_SIZE(sensors); i++) {
```
[the remark this answers is not in the document]
Comment (soburi) [reply samples/sensor/accel_polling/src/main.c]: Fixed it.

File: samples/sensor/accel_polling/src/main.c [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+		if (!device_is_ready(sensors[i])) {
+			printk("sensor: device %s not ready.\n", sensors[i]->name);
+			return;
+		}
+	}
+
+	while (1) {
+		for (uint32_t i = 0; i < ARRAY_SIZE(sensors); i++) {
```
[the remark this answers is not in the document]
Comment (soburi) [reply samples/sensor/accel_polling/src/main.c]: Fixed it.

[the remark this answers is not in the document]
Comment (soburi) [reply samples/sensor/accel_polling/src/main.c]: Thanks for suggestion. Applied it.

File: samples/sensor/accel_polling/src/main.c [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+#include <zephyr/device.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/sys/util_macro.h>
+#include <zephyr/zephyr.h>
+
+#define ACCEL_ALIAS(i) DT_ALIAS(_CONCAT(accel, i))
+#define ACCELEROMETER_DEVICE(i, _)                                                                 \
+	IF_ENABLED(DT_NODE_EXISTS(ACCEL_ALIAS(i)), (DEVICE_DT_GET(ACCEL_ALIAS(i)),))
+
+static const struct device *sensors[] = {LISTIFY(10, ACCELEROMETER_DEVICE, ())};
```
[the remark this answers is not in the document]
Comment (soburi) [reply samples/sensor/accel_polling/src/main.c:20]: Thanks for suggestion. Applied it.

File: samples/sensor/accel_polling/README.rst [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+
+This sample application demonstrates how to use 3-Axis accelerometers.
+
+Building and Running
+********************
+
+This sample requires a 3-Axis accelerometer and needs to set an alias
+that names 'accel0' link to the accelerometer in the devicetree.
+This sample can handle multiple accelerometers by set alias as
+accel1, accel2, ... accel9.
```
[the remark this answers is not in the document]
Comment (soburi) [reply samples/sensor/accel_polling/README.rst]: Thanks for suggestion. Applied it.

File: samples/sensor/accel_polling/sample.yaml [hunk 21]
```diff
@@ -0,0 +1,27 @@
+sample:
+  name: Accelerometer polling sample
+tests:
+  sample.sensor.accel_polling:
+    harness: console
+    tags: sensors
+    filter: dt_alias_exists("accel0")
+    platform_exclude: lpcxpresso55s69_cpu1
```
Comment (gmarull) [inline samples/sensor/accel_polling/sample.yaml:8]: why is this required?

File: samples/sensor/accel_polling/src/main.c [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+	struct sensor_value accel[3];
+
+	ret = sensor_sample_fetch(dev);
+	if (ret < 0) {
+		printk("%s: sensor_sample_fetch() failed: %d\n", dev->name, ret);
+		return ret;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(channels); i++) {
+		ret = sensor_channel_get(dev, channels[i], &accel[i]);
```
Comment (gmarull) [inline samples/sensor/accel_polling/src/main.c:40]: `ret` here is ignored?

Comment (soburi) [reply samples/sensor/accel_polling/src/main.c:40]: Add error check.

File: samples/sensor/accel_polling/sample.yaml (hunk 21 above)
Comment (soburi) [reply samples/sensor/accel_polling/sample.yaml]: I found correct way to modify dts files about lpcxpresso55s69 board.
Remove this line.

**Review by gmarull — APPROVED** [review]

File: boards/arm/96b_wistrio/Kconfig.defconfig [hunk 23]
```diff
@@ -11,4 +11,7 @@ config BOARD
 config REGULATOR
 	default y if LORA
 
+config I2C
+	default y if SENSOR
```
[the remark this answers is not in the document]
Comment (gmarull) [reply boards/arm/96b_wistrio/Kconfig.defconfig:15]: This pattern is not ideal, but still, it's probably good enough for upstream boards and samples as it minimizes the maintenance work. It should probably be documented together with its limitations.

[the remark this answers is not in the document]
Comment (soburi) [reply boards/arm/96b_wistrio/Kconfig.defconfig:15]: > not sure if it is a good idea to default to y as long as SENSOR is enabled, what if user has an overlay to disable the i2c sensor in the default device tree, but has another spi sensor?

My understanding is that the default is should set to what you expect from a usual use case. These boards have sensors on the I2C bus, so this setting should be reasonable.
It may not always be a preferable default value in the complex situations given in the example.
But in such situations, I think it should be explicitly specify configurations, using with prj.conf for examply.

**Review by ycsin — APPROVED** [review]

**Comment by carlescufi** [comment]
@MaureenHelm ping

**Review by carlescufi — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]
This looks great, thank you!
